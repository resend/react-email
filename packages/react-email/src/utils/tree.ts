import { promises as fs } from 'node:fs';
import os from 'node:os';
import path from 'node:path';

const SYMBOLS = {
  BRANCH: '├── ',
  EMPTY: '',
  INDENT: '    ',
  LAST_BRANCH: '└── ',
  VERTICAL: '│   ',
};

const getTreeLines = async (
  dirPath: string,
  depth: number,
  currentDepth = 0,
) => {
  const base = process.cwd();
  const dirFullpath = path.resolve(base, dirPath);
  const dirname = path.basename(dirFullpath);
  let lines = [dirname];

  const dirStat = await fs.stat(dirFullpath);
  if (dirStat.isDirectory() && currentDepth < depth) {
    const childDirents = await fs.readdir(dirFullpath, { withFileTypes: true });

    childDirents.sort((a, b) => {
      // orders directories before files
      if (a.isDirectory() && b.isFile()) {
        return -1;
      }

      if (a.isFile() && b.isDirectory()) {
        return 1;
      }

      // orders by name because they are the same type
      // either directory & directory
      // or file & file
      return b.name > a.name ? -1 : 1;
    });

    for (let i = 0; i < childDirents.length; i++) {
      const dirent = childDirents[i]!;
      const isLast = i === childDirents.length - 1;

      const branchingSymbol = isLast ? SYMBOLS.LAST_BRANCH : SYMBOLS.BRANCH;
      const verticalSymbol = isLast ? SYMBOLS.INDENT : SYMBOLS.VERTICAL;

      if (dirent.isFile()) {
        lines.push(`${branchingSymbol}${dirent.name}`);
      } else {
        const pathToDirectory = path.join(dirFullpath, dirent.name);
        const treeLinesForSubDirectory = await getTreeLines(
          pathToDirectory,
          depth,
          currentDepth + 1,
        );
        lines = lines.concat(
          treeLinesForSubDirectory.map((line, index) =>
            index === 0
              ? `${branchingSymbol}${line}`
              : `${verticalSymbol}${line}`,
          ),
        );
      }
    }
  }

  return lines;
};

export const tree = async (dirPath: string, depth: number) => {
  const lines = await getTreeLines(dirPath, depth);
  return lines.join(os.EOL);
};
