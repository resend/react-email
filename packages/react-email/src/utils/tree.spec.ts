import { tree } from './tree';

test('tree(__dirname, 2)', async () => {
  expect(await tree(__dirname, 2)).toMatchInlineSnapshot(`"utils
├── preview
│   ├── hot-reloading
│   ├── get-env-variables-for-preview-app.ts
│   ├── index.ts
│   ├── serve-static-file.ts
│   └── start-dev-server.ts
├── index.ts
├── tree.spec.ts
└── tree.ts"`);
});
