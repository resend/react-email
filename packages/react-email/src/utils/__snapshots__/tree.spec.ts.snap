// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tree(__dirname, 2) 1`] = `
"utils
├── __snapshots__
│   └── tree.spec.ts.snap
├── esbuild
│   ├── escape-string-for-regex.ts
│   └── renderring-utilities-exporter.ts
├── preview
│   ├── hot-reloading
│   ├── get-env-variables-for-preview-app.ts
│   ├── index.ts
│   ├── serve-static-file.ts
│   └── start-dev-server.ts
├── types
│   ├── hot-reload-change.ts
│   └── hot-reload-event.ts
├── get-emails-directory-metadata.spec.ts
├── get-emails-directory-metadata.ts
├── get-preview-server-location.ts
├── index.ts
├── packageJson.ts
├── register-spinner-autostopping.ts
├── tree.spec.ts
└── tree.ts"
`;
